### PR TITLE
Marginal cost storage

### PR DESCRIPTION
**Pull request by fneum**
## Changes proposed in this Pull Request

Adds marginal cost on storage unit and store energy filling level

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

File: pypsa/optimization/optimize.py [hunk 1]
```diff
@@ -144,6 +144,18 @@ def define_objective(n, sns):
             var = m[f"{c}-{attr}"]
             objective.append((var * cost).sum())
 
+    # marginal cost on energy storage for Stores
+    if not n.stores.empty:
+        e_marginal_cost = n.stores.e_marginal_cost
+        var = m["Store-e"]
+        objective.append((var * e_marginal_cost).sum())
```
Comment (FabianHofmann) [inline pypsa/optimization/optimize.py:151]: just thinking if the marginals costs should be weighted by snapshot weightings?

File: test/test_lopf_storage.py [hunk 2]
```diff
@@ -39,3 +39,29 @@ def network():
 def test_lopf(network, target_gen_p, api):
     optimize(network, api)
     equal(network.generators_t.p.reindex_like(target_gen_p), target_gen_p, decimal=2)
+
+
+@pytest.mark.parametrize("api", ["linopy"])
+def test_storage_energy_marginal_cost(api):
```
Comment (fneum) [inline test/test_lopf_storage.py:45]:
```suggestion
def test_storage_energy_marginal_cost():
```

File: pypsa/optimization/optimize.py (hunk 1 above)
Comment (fneum) [reply pypsa/optimization/optimize.py:151]: Yes, you're right. The cost should be in €/MWh/h, thus requiring a snapshot weighting.

**Comment by fneum** [comment]
xref #1012

**Pull request merged**